### PR TITLE
[FEAT] Handle batches with partial results

### PR DESCRIPTION
**Pull request by MatthieuMoreau0**
- Stop raising an Exception if one of the job of the batch has no results after fetching. 
- Stop raising an exception when fetching results if the status of the batch is a termination status.

Currently if one of the job of a batch fails, it is impossible to retrieve the results for any jobs of the batch - this raises an exception. In this PR, I suggest instead to add `None` to the list of results for that submission when a job has no results. I don't know if returning None is the most idiomatic way to do this in pulser; I also contemplated creating a new subclass of `Result` called `EmptyResult` but I wanted to gather some feedback before going for that because it would imply some refactoring of the `Result` class and I'm not entirely convinced it would bring much.

### CODE REVIEW

**Comment by HGSilveri** [comment]
> I don't know if returning None is the most idiomatic way to do this in pulser; I also contemplated creating a new subclass of `Result` called `EmptyResult` but I wanted to gather some feedback before going for that because it would imply some refactoring of the `Result` class and I'm not entirely convinced it would bring much.

I was thinking of something along those lines too, I think it would be better than `None`. I also considering having `PendingResult` for distinguishing pending jobs from failed ones.

**Comment by HGSilveri** [comment]
I think we both agree that the current format that we use to return results is limiting us and that there doesn't seem to be a nice way to include partial results within it. 
As such, I would propose we add an alternative, more flexible method to `RemoteResults` instead. I was thinking of something along the lines of 
```
RemoteResults.get_available_results() -> dict[JobId, Result]:
```
This would give the user a workaround when something failed in the traditional method. We could also point the user to this alternative method through the error message.
What do you think about this @MatthieuMoreau0 ?

File: pulser-core/pulser/backend/remote.py [hunk 1]
```diff
@@ -139,6 +147,10 @@ def _fetch_result(
         """Fetches the results of a completed submission."""
         pass
 
+    @abstractmethod
+    def _query_result(self, submission_id: str) -> Mapping[str, Result | None]:
```
Comment (HGSilveri) [inline pulser-core/pulser/backend/remote.py:151]: Could you add a short description to this?

File: pulser-core/pulser/backend/remote.py [hunk 2]
```diff
@@ -103,6 +103,14 @@ def get_status(self) -> SubmissionStatus:
         """Gets the status of the remote submission."""
         return self._connection._get_submission_status(self._submission_id)
 
+    def _get_available_result(self, submission_id: str) -> dict[str, Result]:
```
Comment (HGSilveri) [inline pulser-core/pulser/backend/remote.py:106]: Shouldn't we make this public?

File: pulser-core/pulser/backend/remote.py [hunk 3]
```diff
@@ -103,6 +103,14 @@ def get_status(self) -> SubmissionStatus:
...
+        """Return available results and ignore jobs with no results."""
```
Comment (HGSilveri) [inline pulser-core/pulser/backend/remote.py:107]: I think explaining the output format would be helpful

File: pulser-core/pulser/backend/remote.py [hunk 4]
```diff
@@ -103,6 +103,14 @@ def get_status(self) -> SubmissionStatus:
...
+        return {
+            k: v
+            for k, v in self._connection._query_result(submission_id).items()
```
Comment (HGSilveri) [inline pulser-core/pulser/backend/remote.py:110]: You'll want to select the subset of job IDs withint this `RemoteResults` instance

File: pulser-pasqal/pulser_pasqal/pasqal_cloud.py [hunk 5]
```diff
@@ -183,37 +183,44 @@ def _fetch_result(
...
     ) -> tuple[Result, ...]:
         # For now, the results are always sampled results
+        full_results = self._query_result(submission_id)
+
+        if job_ids is None:
+            job_ids = list(full_results.keys())
+
+        results = []
+        for id in job_ids:
+            assert full_results[id] is not None, "Failed to fetch the results."
```
Comment (HGSilveri) [inline pulser-pasqal/pulser_pasqal/pasqal_cloud.py:193]: Let's turn this into a proper `raise` statement, informing the user they can still get the result of successful jobs via the new method

File: pulser-pasqal/pulser_pasqal/pasqal_cloud.py [hunk 6]
```diff
@@ -183,37 +183,44 @@ def _fetch_result(
...
         sdk_jobs = batch.ordered_jobs
-        if job_ids is not None:
-            ind_job_pairs = [
-                (job_ids.index(job.id), job)
-                for job in sdk_jobs
-                if job.id in job_ids
-            ]
-            ind_job_pairs.sort()
-            sdk_jobs = [job for _, job in ind_job_pairs]
+
```
Comment (HGSilveri) [inline pulser-pasqal/pulser_pasqal/pasqal_cloud.py:209]: nit: we don't need `sdk_jobs`, we can give `batch.ordered_jobs` directly to the for loop

**Review by HGSilveri — COMMENTED** [review]
Thanks for your work on this!

File: pulser-pasqal/pulser_pasqal/pasqal_cloud.py (hunk 5 above)
Comment (HGSilveri) [reply pulser-pasqal/pulser_pasqal/pasqal_cloud.py:193]: Let's also make it conditional on the status?

File: pulser-core/pulser/backend/remote.py (hunk 2 above)
Comment (HGSilveri) [reply pulser-core/pulser/backend/remote.py:106]:
```suggestion
    def get_available_results(self, submission_id: str) -> dict[str, Result]:
```

File: pulser-core/pulser/backend/remote.py (hunk 1 above)
Comment (HGSilveri) [reply pulser-core/pulser/backend/remote.py:151]: nit: Just a bit clear to me
```suggestion
    def _query_job_results(self, submission_id: str) -> Mapping[str, Result | None]:
```

File: pulser-core/pulser/backend/remote.py [hunk 7]
```diff
@@ -103,6 +103,14 @@ def get_status(self) -> SubmissionStatus:
...
+            if v is not None
+        }
+
     def __getattr__(self, name: str) -> Any:
```
Comment (HGSilveri) [inline pulser-core/pulser/backend/remote.py:114]: Let's stop blocking the call to `_fetch_result()` depending on the status

File: pulser-core/pulser/backend/remote.py (hunk 1 above)
Comment (MatthieuMoreau0) [reply pulser-core/pulser/backend/remote.py:151]: Done.

fetch_result uses singular while query_job_results is now plural. Is that ok?

File: pulser-core/pulser/backend/remote.py (hunk 2 above)
Comment (MatthieuMoreau0) [reply pulser-core/pulser/backend/remote.py:106]: Done

File: pulser-core/pulser/backend/remote.py (hunk 3 above)
Comment (MatthieuMoreau0) [reply pulser-core/pulser/backend/remote.py:107]: Added some extra comment, lmk if that looks good to you

File: pulser-core/pulser/backend/remote.py (hunk 4 above)
Comment (MatthieuMoreau0) [reply pulser-core/pulser/backend/remote.py]: Done

File: pulser-pasqal/pulser_pasqal/pasqal_cloud.py (hunk 5 above)
Comment (MatthieuMoreau0) [reply pulser-pasqal/pulser_pasqal/pasqal_cloud.py]: Should be good

File: pulser-core/pulser/backend/remote.py (hunk 1 above)
Comment (HGSilveri) [reply pulser-core/pulser/backend/remote.py]: Yes, let's not overthink it :)

File: pulser-pasqal/pulser_pasqal/pasqal_cloud.py [hunk 8]
```diff
@@ -183,19 +185,27 @@ def _fetch_result(
...
-            job_ids = list(full_results.keys())
+            job_ids = list(jobs.keys())
 
-        results = []
+        results: list[Result] = []
         for id in job_ids:
-            assert full_results[id] is not None, "Failed to fetch the results."
-            results.append(cast(Result, full_results[id]))
+            status, result = jobs[id]
+            if status != {JobStatus.PENDING, JobStatus.RUNNING}:
```
Comment (HGSilveri) [inline pulser-pasqal/pulser_pasqal/pasqal_cloud.py:196]: Didn't you mean
```suggestion
            if status in {JobStatus.PENDING, JobStatus.RUNNING}:
```
?

File: pulser-core/pulser/backend/remote.py [hunk 9]
```diff
@@ -148,7 +173,16 @@ def _fetch_result(
         pass
 
     @abstractmethod
-    def _query_result(self, submission_id: str) -> Mapping[str, Result | None]:
+    def _query_job_progress(
+        self, submission_id: str
+    ) -> Mapping[str, tuple[JobStatus, Result | None]]:
+        """Fetches the status and results of a submission.
```
Comment (HGSilveri) [inline pulser-core/pulser/backend/remote.py:179]: nit: 
```suggestion
        """Fetches the status and results of all the jobs in a submission.
```

File: pulser-pasqal/pulser_pasqal/pasqal_cloud.py (hunk 5 above)
Comment (HGSilveri) [reply pulser-pasqal/pulser_pasqal/pasqal_cloud.py]: Yup!

File: tests/test_pasqal.py [hunk 10]
```diff
@@ -77,40 +82,38 @@ def __init__(
...
+            _MockJob(),
             _MockJob(result={"00": 10}),
             _MockJob(result={"11": 10}),
         ]
-        sequence_builder = seq_.to_abstract_repr()
+    )
+    seq_ = copy.deepcopy(_seq)
+    seq_.declare_channel("rydberg_global", "rydberg_global")
+    seq_.measure()
+    sequence_builder = seq_.to_abstract_repr()
```
Comment (HGSilveri) [inline tests/test_pasqal.py:108]: It's weird to have this within the class, why did you need it?

File: tests/test_pasqal.py [hunk 11]
```diff
@@ -77,40 +82,38 @@ def __init__(
...
-        status = "DONE"
-        ordered_jobs = [
-            mock_job,
+@dataclasses.dataclass
+class MockBatch:
+    id = "abcd"
+    status: str = "DONE"
+    ordered_jobs: list[_MockJob] = dataclasses.field(
+        default_factory=lambda: [
+            _MockJob(),
```
Comment (HGSilveri) [inline tests/test_pasqal.py:100]: If I get it right, the changes were needed so that you gained access to take the `MockBatch` definition outside of the fixture?

File: tests/test_pasqal.py [hunk 12]
```diff
@@ -190,15 +196,82 @@ def test_remote_results(fixt, mock_batch, with_job_id):
...
+
+    fixt = mock_pasqal_cloud_sdk(batch)
+
+    remote_results = RemoteResults(
+        batch.id,
+        fixt.pasqal_cloud,
+    )
+
+    fixt.mock_cloud_sdk.get_batch.reset_mock()
+    with pytest.raises(RemoteResultsError):
```
Comment (HGSilveri) [inline tests/test_pasqal.py:228]: Can you check the error message matches what you expect?

File: tests/test_pasqal.py [hunk 13]
```diff
@@ -190,15 +196,82 @@ def test_remote_results(fixt, mock_batch, with_job_id):
...
+        ],
+    )
+
+    fixt = mock_pasqal_cloud_sdk(batch)
+    remote_results = RemoteResults(
+        batch.id,
+        fixt.pasqal_cloud,
+    )
+
+    with pytest.raises(RemoteResultsError):
```
Comment (HGSilveri) [inline tests/test_pasqal.py:255]: Same here, the message should be different right?

File: tests/test_pasqal.py [hunk 14]
```diff
@@ -190,15 +196,82 @@ def test_remote_results(fixt, mock_batch, with_job_id):
...
+    assert available_results == {
+        job.id: SampledResult(
+            atom_order=("q0", "q1", "q2", "q3"),
+            meas_basis="ground-rydberg",
+            bitstring_counts=job.result,
+        )
+        for job in batch.ordered_jobs
+        if job.result is not None
+    }
+
```
Comment (HGSilveri) [inline tests/test_pasqal.py:267]: If you could, some checks that the expected methods in the mock are being called would be nice

**Review by HGSilveri — COMMENTED** [review]
I think I got why you needed so many changes in the tests but let me know if I got it wrong

File: tests/test_pasqal.py (hunk 11 above)
Comment (MatthieuMoreau0) [reply tests/test_pasqal.py:100]: Yeah the goal was to be able to customize the MockBatch, e.g. to set its status or results. This enables testing for the "partial results" case

File: tests/test_pasqal.py (hunk 12 above)
Comment (MatthieuMoreau0) [reply tests/test_pasqal.py:228]: Ah yes I very much intended to do that, on it!

File: tests/test_pasqal.py (hunk 13 above)
Comment (MatthieuMoreau0) [reply tests/test_pasqal.py:255]: Ah in fact the message we catch in the test is the same because we catch the error and re-raise it [here](https://github.com/pasqal-io/Pulser/blob/mm/partial-results/pulser-core/pulser/backend/remote.py#L150). I did this because it felt wrong to mention the `get_vailable_results` method of the `RemoteResults` object inside the `PasqalCloud` class.

The details of why the results are not available will be displayed in the stacktrace though. If you think having to check the stacktrace is unpractical, I can maybe concatenate the error messages instead.

Comment (HGSilveri) [reply tests/test_pasqal.py:255]: That sounds like overkill, no need to bother

File: tests/test_pasqal.py (hunk 10 above)
Comment (MatthieuMoreau0) [reply tests/test_pasqal.py:108]: It was in the fixture before. I can move it to the top level if you prefer

Note: I'm not sure we even need a valid sequence builder, I'll remove it and see if tests still pass

File: tests/test_pasqal.py (hunk 14 above)
Comment (MatthieuMoreau0) [reply tests/test_pasqal.py:297]: Done

File: tests/test_pasqal.py (hunk 10 above)
Comment (MatthieuMoreau0) [reply tests/test_pasqal.py]: I moved it to another function, let me know if that's better for you

Comment (HGSilveri) [reply tests/test_pasqal.py]: Yeah, looks fine :)
